File: src/datalevin/native_value.clj
(ns ^:no-doc datalevin.native-value
  "Runtime bindings for native values in disposable, process-local spill stores."
  (:require [taoensso.nippy :as nippy])
  (:import [datalevin NativeValue]
           [java.io DataInput DataOutput]
           [java.util HashMap]))

(def ^:dynamic *spill-bindings* nil)

(defn spill-bindings
  "Create bindings owned by one spill collection, retaining no value payloads."
  ^HashMap []
  (HashMap.))

(nippy/extend-freeze NativeValue :datalevin/native-spill
  [^NativeValue value ^DataOutput out]
  (when-not *spill-bindings*
    (throw (ex-info "Cannot serialize a native value outside its spill store"
                    {:type-name (.typeName value)})))
  (let [id       (.codecId value)
        bindings ^HashMap *spill-bindings*]
    (when-not (.containsKey bindings id)
      (.put bindings id (.withPayload value (byte-array 0))))
    (.writeUTF out id)
    (nippy/freeze-to-out! out (.payload value))))

(nippy/extend-thaw :datalevin/native-spill
  [^DataInput in]
  (let [id (.readUTF in)]
    (if-let [^NativeValue binding
             (when *spill-bindings* (.get ^HashMap *spill-bindings* id))]
      (.withPayload binding (nippy/thaw-from-in! in))
      (throw (ex-info "Missing native value binding for spill store"
                      {:codec-id id})))))

File: src/java/datalevin/NativeValue.java
package datalevin;

import clojure.lang.IHashEq;

import java.util.Objects;
import java.util.function.BiPredicate;

/**
 * An immutable byte snapshot of a value owned by an embedded language runtime.
 *
 * <p>The bridge supplies logical equality over decoded values. Serialized byte
 * equality is insufficient: equal values can have different payloads. A constant
 * hash keeps that contract even for host values without a stable hash function.
 * This object is a runtime bridge value, not a persistent or wire codec.
 */
public final class NativeValue implements IHashEq {
    private final String codecId;
    private final String typeName;
    private final byte[] payload;
    private final BiPredicate<NativeValue, NativeValue> equality;

    public NativeValue(String codecId, String typeName, byte[] payload,
                       BiPredicate<NativeValue, NativeValue> equality) {
        this.codecId = Objects.requireNonNull(codecId, "codecId");
        this.typeName = Objects.requireNonNull(typeName, "typeName");
        this.payload = Objects.requireNonNull(payload, "payload").clone();
        this.equality = Objects.requireNonNull(equality, "equality");
    }

    public String codecId() {
        return codecId;
    }

    public String typeName() {
        return typeName;
    }

    public byte[] payload() {
        return payload.clone();
    }

    /** Restore a spill snapshot using this value's live runtime binding. */
    public NativeValue withPayload(byte[] bytes) {
        return new NativeValue(codecId, typeName, bytes, equality);
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) {
            return true;
        }
        return other instanceof NativeValue value
                && equality.test(this, value);
    }

    @Override
    public int hashCode() {
        return 0;
    }

    @Override
    public int hasheq() {
        return hashCode();
    }

    @Override
    public String toString() {
        return "#<native-value " + typeName + ">";
    }
}
